### PR TITLE
JavaScript Fundamentals 5: Give better example of reduce, map and filter methods.

### PR DESCRIPTION
**Pull request by nikitarevenco**
## Because

When I first did this lesson, I remember being confused about these three functions. I didn't fully understand just how _powerful_ they are. 

I think that in its current form, the lesson doesn't clearly give a concrete example of why someone would want to use these functions. Yes, the video provided does list some usecases, but I believe that adding these changes will help solidify learner's understanding. (_also I really like the image_)

## This PR

- Add a code example that doesn't use these methods, which makes it more complex 
- Add a code example for `.reduce`, `.map` and `.filter` methods to help learners see the differences
- Add an image to represent this function visually
- Place the material regarding these functions in the assignment section

## Issue

Closes #XXXXX

## Pull Request Requirements

-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [x] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [x] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:44]: Just for increased clarity, let's add a note that this function introduces three problems that need to be solved. Can you think of what they are? (Possibly a spoiler tag)

1. We need to perform an operation on *only the even numbers*
1. We need to transform those numbers into triple their value
1. We need to add the results up

This can help to clarify why we might need to call three different methods to get our result.

Additionally, this can give us a chance to explain filter, map, and reduce. Filter: I need an array with some of these items, but possibly not all of them. Map: I need an array with all of these items, but the values need to be transformed in some way. Reduce: I need some data (any type) that can only be calculated with all of the values in this array.

**Review by fortypercenttitanium — CHANGES_REQUESTED** [review]
Great stuff!

- I'd like to see just a small change for increased clarity.
- This would also be a good place to provide some clarity on callbacks. If you can find an elegant way to include this is well that would be awesome, but not required for a merge.
- Notice the github actions linter pointing out some markdown that needs fixing for the lesson to meet our style guidelines.

Thank you!

File: foundations/javascript_basics/fundamentals-5.md
Comment (nikitarevenco) [reply foundations/javascript_basics/fundamentals-5.md]: I've completely overhauled the section, now it goes in-depth about each method specifically and also gives a brief introduction on callbacks. What do you think about it?

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:90]:
```suggestion
Now, instead of passing a function that adds 1, let's pass a function that multiplies by 3:
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:95]:
```suggestion
#### The filter method
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:66]:
```suggestion
#### The map method
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:68]:
```suggestion
`map` is one such function. It expects a `callback` as an argument, which is a fancy way to say "I want you to pass another function as an argument to my function".
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:71]:
```suggestion
And let's say we had an array of numbers, `const arr = [1, 2, 3, 4, 5]` and we'd like to increment all of these numbers by 1 using our `addOne` function.
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:70]:
```suggestion
Let's say we had a function `addOne`, which takes in `num` as an argument and outputs that `num` increased by 1. 
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:76]:
```suggestion
  return num + 1;
```
Keep it simple and don't mutate

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:72]:
```suggestion
Instead of making a `for` loop and iterating over the above array, we could use our `map` array method instead, which **automatically** iterates over an array for us. 
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:100]:
```suggestion
The `filter` method expects the `callback` to return either `true` or `false`. If it returns `true`, the value is included in the output. Otherwise, it isn't.
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:108]:
```suggestion
- `filter` will iterate through `arr` and pass **every value** into the `isOdd` callback function, one at a time. 
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:109]:
```suggestion
- `isOdd` will return `true` when the value is odd, which means this value is included in the output.
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:110]:
```suggestion
- If it's an even number, `isOdd` will return `false` and not include it in the final output.
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:112]:
```suggestion
#### The reduce method
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:115]:
```suggestion
First, we'd have to declare a variable `total` and initialize it to 1. Then, we'd iterate through the array with a `for` loop and multiply the `total` by the current number.
```
If you initialized to 0, the result would always be 0 😄

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:117]:
```suggestion
But we don't actually need to do all of that, we have our `reduce` method that will do the job for us. Just like `.map()` and `.reduce()` it expects a callback function.
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:118]:
```suggestion
However, there are two key differences with this array method:
- The callback function takes two arguments instead of one. The first argument is the `accumulator`, which is the current value of the result *at that point in the loop*. The first time through, this. value will either be set to the `initialValue` (described in the next bullet), or the first element in the array if no `initialValue` is provided. The second argument for the callback is the `current` value, which is the item currently being iterated on.
- It also takes in an `initialValue` as a second argument (after the callback), which helps when we don't want our initial value to be the first element in the array. For instance, if we wanted to sum all numbers in an array, we could call reduce without an `initialValue`, but if we wanted to sum all numbers in an array and add 10, we could use 10 as our `initialValue`.
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md]:
```suggestion
So `reduce` will, like the other methods, go through every element in `arr` and apply the `callback` function to it. It then **stores** the output in `total`, which in the previous example was 1. 
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:126]:
```suggestion
- Initialize total to `1` in the second argument.
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:132]:
```suggestion
You've learnt about the three powerful array methods which are `map`, `filter` and `reduce`. They allow us to write shorter code that is more readable and less prone to bugs. 
```

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:138]:
```suggestion
Let's do some quick practice before your assignment! Rewrite the `sumOfTripledEvens(array)` function using these three methods.
```
I don't think that hint is particularly helpful, they will see once they click the result.

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:140]: Once you are finished and you've tested that your function works correctly, check out the solution below:
<details>
  <summary>Solution</summary>
  <pre><code lang="js">
  function sumOfTripledEvens(array) {
    return array
      .filter(num => num % 2 === 0)
      .map(num => num * 3)
      .reduce((acc, curr) => acc + curr);
  }
  </code></pre>
</details>

Comment (fortypercenttitanium) [reply foundations/javascript_basics/fundamentals-5.md:140]: The above is another suggestion, but we'll need to make sure the formatting is correct

**Review by fortypercenttitanium — CHANGES_REQUESTED** [review]
Awesome work. Just a few more changes, some stylistic and some of my own content that I think improves on what you're authoring here. This lesson is going to be much better after this gets merged.

File: foundations/javascript_basics/fundamentals-5.md
Comment (nikitarevenco) [reply foundations/javascript_basics/fundamentals-5.md:137]: How about: 'example of filter, map and reduce methods being used to visually represent making a sandwich'. 

Let me know what you think about this. I think the image is worth including because it is a nice visual representation of what we've been talking about in the lesson and people will find it helpful.

Comment (thatblindgeye) [reply foundations/javascript_basics/fundamentals-5.md:137]: That sounds great to me!

**Review by thatblindgeye — APPROVED** [review]
Approving for my comment above, will defer to @fortypercenttitanium for the remainder

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:148]: I don't think the backticks inside the details block is going to work, at least according to the [lesson preview tool](https://www.theodinproject.com/lessons/preview). My suggestion of using the code elements might though. @thatblindgeye do you have any suggestions regarding this?

File: foundations/javascript_basics/fundamentals-5.md
Comment (fortypercenttitanium) [inline foundations/javascript_basics/fundamentals-5.md:34]:
```suggestion
1. We need to transform *those* numbers by multiplying them by 3.
```
More explicit. "Triple numbers" is somewhat vague

**Review by fortypercenttitanium — COMMENTED** [review]
Check comments

File: foundations/javascript_basics/fundamentals-5.md
Comment (JoshDevHub) [reply foundations/javascript_basics/fundamentals-5.md:148]: You need to give the `details` element the attribute `markdown="block"`
```md
<details markdown="block">
</details>
```
Then it will render the markdown properly.

Comment (nikitarevenco) [reply foundations/javascript_basics/fundamentals-5.md]: I've added 

> You need to give the `details` element the attribute `markdown="block"`
> 
> ```
> <details markdown="block">
> </details>
> ```
> 
> Then it will render the markdown properly.

Added it, awesome. It works correctly now.

**Review by fortypercenttitanium — APPROVED** [review]
LGTM!

**Review by fortypercenttitanium — APPROVED** [review]

**Pull request merged**
